File: apps/web/src/model/ops/transform.ts
import type { ArrayLayout } from '../../contracts/generated/ArrayLayout';
import type { Transform } from '../../contracts/generated/Transform';
import type { Entity, EntityKind, NewEntity } from '../entities';
import type { Affine } from '../geom/affine';
import { arrayObjects as coreArrayObjects, transformObjects as coreTransformObjects } from '../../wasm/core';
import { packEntities, unpackEntities, type Geometry, type Packed } from '../../wasm/pack';
import { entityOp } from './entityOp';

/**
 * Applies a similarity transform (move, rotate, uniform scale, mirror) to
 * any entity. Returns a copy with the same id; callers decide whether to
 * update the original or add the copy. Computed by the geometry core
 * (docs/adr/0008).
 */
export const transformEntity = entityOp<<E extends Entity>(e: E, m: Affine) => E>('transformEntity');

export const translateEntity = entityOp<<E extends Entity>(e: E, dx: number, dy: number) => E>('translateEntity');

/**
 * `transformEntity` of every entity by every affine, affine after affine,
 * in one call, through JSON. The tools ask a geometry store instead, which
 * holds the objects already (`transformedFrom`); this JSON call stays as
 * the reference that path is held to (src/wasm/transform.wasm.test.ts).
 */
export const transformEntities = entityOp<<E extends Entity>(list: readonly E[], ms: readonly Affine[]) => E[]>('transformEntities');

/** A kind's geometry fields, as the core's JSON names them (geometry-core entity.rs `Shape`). */
export const SHAPE_FIELDS: Record<EntityKind, readonly string[]> = {
  point: ['p', 'z'],
  line: ['a', 'b'],
  polyline: ['pts', 'bulges', 'holes'],
  polygon: ['pts', 'bulges', 'holes'],
  circle: ['c', 'r'],
  arc: ['c', 'r', 'a0', 'a1'],
  ellipse: ['c', 'major', 'ratio', 't0', 't1'],
  xline: ['p', 'dir'],
  ray: ['p', 'dir'],
  spline: ['pts', 'closed'],
  text: ['p', 'text', 'height', 'rotation'],
  dimension: ['a', 'b', 'offset', 'height', 'text', 'style', 'angle', 'c'],
  hatch: ['ring', 'holes', 'pattern'],
};

/**
 * Every number in an object's geometry fields is finite: its points, radii,
 * angles, bulges, heights, offsets, a hatch pattern's angle and spacing.
 * `cad.entities.transform` refuses a transform that carries one past the
 * largest float64 (docs/adr/0037); the desktop checks the same numbers.
 */
export function geometryIsFinite(e: Entity): boolean {
  const ok = (v: unknown): boolean => (typeof v === 'number' ? Number.isFinite(v) : Array.isArray(v) ? v.every(ok) : v !== null && typeof v === 'object' ? Object.values(v).every(ok) : true);
  const src = e as unknown as Record<string, unknown>;
  return (SHAPE_FIELDS[e.kind] ?? []).every((key) => ok(src[key]));
}

/**
 * `e` with another geometry, laid out as the JSON call gives it back: every
 * other field of its own first, then the kind and its fields. The other
 * fields keep their values; the attributes, the one object among them,
 * are copied, so a copy shares nothing with its original (as after JSON).
 * A cleared `bulges` or `holes` is written as undefined, as `entityOp`
 * does, so `CadDocument.update` drops the old one.
 */
function withGeometry<E extends Entity | NewEntity>(e: E, g: Geometry): E {
  const src = e as unknown as Record<string, unknown>;
  const own = SHAPE_FIELDS[src.kind as EntityKind] ?? [];
  const out: Record<string, unknown> = {};
  for (const key in src) {
    if (key === 'kind' || own.includes(key)) continue;
    const v = src[key];
    out[key] = key === 'attrs' && v && typeof v === 'object' ? { ...v } : v;
  }
  for (const key in g) out[key] = g[key];
  if ((g.kind === 'polyline' || g.kind === 'polygon') && !('bulges' in g)) out.bulges = undefined;
  if ((g.kind === 'polygon' || g.kind === 'hatch') && !('holes' in g)) out.holes = undefined;
  return out as unknown as E;
}

/**
 * `transformEntities(list, ms)` read from a geometry store's packed answer
 * (`CoreStore.transformPacked` over `ids`, the list's objects as that store
 * numbers them, and `count` affines; null: as many as the answer holds, an
 * array's copies): the store transforms its own copies and only the new
 * geometry crosses, as numbers (docs/adr/0008). Each object keeps its other
 * fields; the result is the JSON call's, field for field, except that −0
 * survives here as everywhere the store is packed.
 */
export function transformedFrom<E extends Entity | NewEntity>(list: readonly E[], ids: ArrayLike<number>, count: number | null, packed: Packed): E[] {
  const records = unpackEntities(packed);
  const out: E[] = [];
  let at = 0;
  const runs = count ?? records.length / list.length;
  for (let k = 0; k < runs; k++)
    for (let i = 0; i < list.length; i++) {
      const r = records[at++];
      // The store skips an id it does not hold: the drawing and its copy would have drifted apart.
      if (!r || r.id !== ids[i]) throw new Error('Geometri deposu seçili nesnelerin hepsini bulamadı; işlem uygulanmadı. Sayfayı yenileyip yeniden deneyin.');
      out.push(withGeometry(list[i], r.geometry));
    }
  if (at !== records.length) throw new Error('Geometri deposu beklenenden çok nesne döndürdü; işlem uygulanmadı.');
  return out;
}

/** A transform of `cad.entities.transform` as the core's `similarity` takes it: its kind and numbers. */
function similarityOf(t: Transform): [string, number[]] {
  switch (t.kind) {
    case 'move':
      return ['move', [t.dx, t.dy]];
    case 'rotate':
      return ['rotate', [t.center.x, t.center.y, t.angle]];
    case 'scale':
      return ['scale', [t.center.x, t.center.y, t.factor]];
    case 'mirror':
      return ['mirror', [t.a.x, t.a.y, t.b.x, t.b.y]];
    case 'align': {
      const first = [t.source.x, t.source.y, t.target.x, t.target.y];
      if (!t.source2 || !t.target2) return ['align', first];
      return [t.scale === true ? 'alignScale' : 'align', [...first, t.source2.x, t.source2.y, t.target2.x, t.target2.y]];
    }
  }
}

/**
 * `list` moved by one transform of the product command
 * `cad.entities.transform` (docs/adr/0037), with no store and no JSON: the
 * objects are packed (../../wasm/pack.ts), the core builds the matrix from
 * the transform's numbers and moves every object (`transform_packed_objects`),
 * and only the new geometry comes back. Each object keeps its other fields,
 * as `transformedFrom` keeps them; −0 and every other float64 survive, which
 * the JSON call (`transformEntities`) cannot promise.
 */
export function transformObjects<E extends Entity>(list: readonly E[], t: Transform): E[] {
  if (!list.length) return [];
  const packed = packEntities(list);
  const [kind, params] = similarityOf(t);
  const moved = coreTransformObjects(packed.nums, packed.strings, kind, Float64Array.from(params));
  return transformedFrom(
    list,
    list.map((e) => e.id),
    1,
    moved,
  );
}

/** An array of `cad.entities.array` as the core's `array_transforms` takes it: its kind and numbers. */
export function arrayNumbers(layout: ArrayLayout): [string, number[]] {
  if (layout.kind === 'grid') return ['grid', [layout.rows, layout.cols, layout.dx, layout.dy]];
  return ['polar', [layout.center.x, layout.center.y, layout.count, layout.fill, layout.rotate ? 1 : 0]];
}

/**
 * Copies of `list` laid out by an array of the product command
 * `cad.entities.array` (docs/adr/0047), with no store and no JSON: the core
 * lays the copies out (`array_transforms`; a polar array that does not turn
 * places them by the middle of the list's box, text measured in `font`, a
 * `DrawingFont` id) and moves every object, packed. The copies come place
 * after place, each place in the list's order, each with its original's
 * other fields.
 */
export function arrayCopies<E extends Entity>(list: readonly E[], layout: ArrayLayout, font: string): E[] {
  if (!list.length) return [];
  const packed = packEntities(list);
  const [kind, params] = arrayNumbers(layout);
  const copies = coreArrayObjects(packed.nums, packed.strings, kind, Float64Array.from(params), font);
  // As many runs of the list as the core laid places out (rows × cols − 1, or count − 1).
  return transformedFrom(
    list,
    list.map((e) => e.id),
    null,
    copies,
  );
}

File: crates/native/application/src/transform.rs
//! `cad.entities.transform` v1 (docs/adr/0037): objects named by their
//! persistent ids moved, rotated, scaled or mirrored as one undo step, in
//! place or as copies. The desktop's handler over the native document; the
//! web's is `apps/web/src/product/entitiesTransform.ts`. Both pass the shared
//! cases in `fixtures/commands/v1/cad.entities.transform.json`.
//!
//! The modify tools (Taşı, Kopyala, Döndür, Ölçekle, Aynala) make the
//! selection explicit here (TODOS.md CMD-07): they give the selected
//! objects' ids. The geometry is the shared core's, as on the web: the
//! matrix of the transform (`similarity`) and what it does to each kind of
//! object (`transform_shape`); nothing is computed here.
//!
//! The checks, in order (the first that fails answers):
//! 1. at least one id; every id lowercase UUID text with hyphens (in order);
//! 2. the transform's numbers finite, in their order; a scale factor above
//!    zero; a mirror axis with a direction;
//! 3. the expected revision (every command's, `checks.rs`);
//! 4. every id names an object of the document (in order);
//! 5. not every object on a locked layer (with others, a warning);
//! 6. no coordinate carried past the largest float64 by the transform.
//!
//! Objects on a locked layer are neither changed nor copied. The web's tools
//! used to copy them onto their locked layer; ADR 0037 records the change.

use kentos_contracts::{
    CommandError, CommandResult, CommandWarning, EntitiesTransform, EntitiesTransformPlan,
    EntitiesTransformed, Entity, Transform,
};
use kentos_domain::{Document, Slot};
use kentos_geometry_core::entity::Shape;
use kentos_geometry_core::geom::affine::{Affine, similarity};
use kentos_geometry_core::ops::transform::transform_shape;

use crate::ExecutionContext;
use crate::checks::{self, Stop};
/// The stable codes of the answers (`CommandError.code`, `CommandWarning.code`).
pub use crate::codes;
use crate::geometry::{shape, with_shape};

/// Checks `input` against the document, writing nothing.
pub fn validate(cx: &ExecutionContext<'_>, input: &EntitiesTransform) -> CommandResult<()> {
    match check(cx.doc, input) {
        Ok(checked) => CommandResult::Completed {
            output: (),
            warnings: checked.warnings,
        },
        Err(stop) => stop.into(),
    }
}

/// What execute would write now: each object as it would be, and the
/// revision to expect for exactly that; writes nothing.
pub fn plan(
    cx: &ExecutionContext<'_>,
    input: &EntitiesTransform,
) -> CommandResult<EntitiesTransformPlan> {
    let copy = input.copy.unwrap_or(false);
    match check(cx.doc, input) {
        Ok(checked) => CommandResult::Completed {
            output: EntitiesTransformPlan {
                sources: checked.sources.iter().map(|(_, uid)| uid.clone()).collect(),
                entities: checked
                    .moved
                    .into_iter()
                    .map(|mut e| {
                        // A copy's slot is given when it is written.
                        if copy {
                            base_mut(&mut e).id = 0;
                        }
                        e
                    })
                    .collect(),
                locked: checked.locked,
                revision: cx.doc.revision().to_string(),
            },
            warnings: checked.warnings,
        },
        Err(stop) => stop.into(),
    }
}

/// Checks `input` again and writes it as one undo step named after the tool:
/// the objects changed in place, or their copies added. Into the open
/// transaction or group, if one is.
pub fn execute(
    cx: &mut ExecutionContext<'_>,
    input: EntitiesTransform,
) -> CommandResult<EntitiesTransformed> {
    let copy = input.copy.unwrap_or(false);
    let checked = match check(cx.doc, &input) {
        Ok(checked) => checked,
        Err(stop) => return stop.into(),
    };
    let label = label(&input.transform, copy);
    let (changed, created) = if copy {
        match cx.doc.add_many(checked.moved, label) {
            Ok(slots) => {
                let doc = &*cx.doc;
                let created = slots
                    .iter()
                    .filter_map(|slot| doc.uid(*slot))
                    .map(|uid| uid.to_string())
                    .collect();
                (Vec::new(), created)
            }
            Err(full) => {
                return CommandResult::Failed {
                    error: CommandError {
                        code: codes::SLOTS_EXHAUSTED.into(),
                        message: full.to_string(),
                        path: None,
                        revision: None,
                    },
                };
            }
        }
    } else {
        let changes: Vec<(Slot, Entity)> = checked
            .sources
            .iter()
            .map(|(slot, _)| *slot)
            .zip(checked.moved)
            .collect();
        cx.doc.update_many(changes, label);
        let changed = checked.sources.into_iter().map(|(_, uid)| uid).collect();
        (changed, Vec::new())
    };
    CommandResult::Completed {
        output: EntitiesTransformed {
            changed,
            created,
            locked: checked.locked,
            revision: cx.doc.revision().to_string(),
        },
        warnings: checked.warnings,
    }
}

/// The undo step's name: the tool's (docs/adr/0037).
pub fn label(transform: &Transform, copy: bool) -> &'static str {
    match transform {
        Transform::Move { .. } if copy => "Kopyala",
        Transform::Move { .. } => "Taşı",
        Transform::Rotate { .. } => "Döndür",
        Transform::Scale { .. } => "Ölçekle",
        Transform::Mirror { .. } => "Aynala",
    }
}

/// The affine of a transform, built by the shared core (`similarity`), as
/// the web's handler builds it through WASM. None only for numbers the
/// checks would refuse first.
pub fn affine(transform: &Transform) -> Option<Affine> {
    match *transform {
        Transform::Move { dx, dy } => similarity("move", &[dx, dy]),
        Transform::Rotate { center, angle } => similarity("rotate", &[center.x, center.y, angle]),
        Transform::Scale { center, factor } => similarity("scale", &[center.x, center.y, factor]),
        Transform::Mirror { a, b } => similarity("mirror", &[a.x, a.y, b.x, b.y]),
    }
}

/// What may be written: the objects to transform (slot and id) and each as
/// it would be, those that stay on locked layers, and the warning about them.
struct Checked {
    sources: Vec<(Slot, String)>,
    moved: Vec<Entity>,
    locked: Vec<String>,
    warnings: Vec<CommandWarning>,
}

fn locked_message(n: usize) -> String {
    format!(
        "{n} nesne kilitli katmanda olduğu için atlandı. Değiştirmek için katmanın kilidini Katmanlar panelinden açın."
    )
}

/// The transform's own checks, in its fields' order: finite numbers, then
/// what they mean. Its affine when they pass.
fn check_transform(transform: &Transform) -> Result<Affine, Stop> {
    let finite =
        |value: f64, what: &str, fix: &str, path: &str| checks::finite(value, what, fix, path);
    match *transform {
        Transform::Move { dx, dy } => {
            let fix = "Kaydırmayı sonlu bir sayıyla verin.";
            finite(dx, "Doğu (Y) yönündeki kaydırma", fix, "transform.dx")?;
            finite(dy, "Kuzey (X) yönündeki kaydırma", fix, "transform.dy")?;
        }
        Transform::Rotate { center, angle } => {
            checks::point(center, "Merkezin", "transform.center")?;
            finite(
                angle,
                "Dönme açısı",
                "Açıyı sonlu bir sayıyla verin.",
                "transform.angle",
            )?;
        }
        Transform::Scale { center, factor } => {
            checks::point(center, "Merkezin", "transform.center")?;
            finite(
                factor,
                "Ölçek faktörü",
                "Faktörü sonlu bir sayıyla verin.",
                "transform.factor",
            )?;
            if factor <= 0.0 {
                return Err(Stop::Failed(checks::error(
                    codes::INVALID_FACTOR,
                    "Ölçek faktörü sıfırdan büyük olmalı. Pozitif bir faktör verin.".into(),
                    Some("transform.factor".into()),
                )));
            }
        }
        Transform::Mirror { a, b } => {
            checks::point(a, "Eksenin ilk noktasının", "transform.a")?;
            checks::point(b, "Eksenin ikinci noktasının", "transform.b")?;
            // The core's own measure of the axis (`mirror`): zero leaves it no direction.
            let (dx, dy) = (b.x - a.x, b.y - a.y);
            if dx * dx + dy * dy == 0.0 {
                return Err(Stop::Failed(checks::error(
                    codes::INVALID_AXIS,
                    "Simetri ekseninin iki noktası aynı; eksenin yönü yok. Birbirinden ayrı iki nokta verin.".into(),
                    Some("transform.b".into()),
                )));
            }
        }
    }
    affine(transform).ok_or_else(|| {
        Stop::Failed(checks::error(
            codes::NOT_FINITE,
            "Dönüşüm kurulamadı.".into(),
            Some("transform".into()),
        ))
    })
}

/// The checks in the contract's order.
fn check(doc: &Document, input: &EntitiesTransform) -> Result<Checked, Stop> {
    checks::uids(&input.uids, "Dönüştürülecek nesne verilmedi.")?;
    let m = check_transform(&input.transform)?;
    checks::revision(doc, input.expected_revision.as_deref())?;
    let mut sources = Vec::new();
    let mut moved = Vec::new();
    let mut locked = Vec::new();
    let mut overflow = false;
    for (slot, entity, uid) in checks::objects(doc, &input.uids)? {
        if doc.layers().is_locked(&entity.base().layer_id) {
            locked.push(uid.clone());
            continue;
        }
        let before = shape(entity);
        let after = transform_shape(&before, &m);
        overflow |= finite_shape(&before) && !finite_shape(&after);
        // The core keeps an object's kind; a shape of another kind would be its fault, not the input's.
        let Some(e) = with_shape(entity, after) else {
            return Err(Stop::Failed(checks::error(
                codes::NOT_FINITE,
                "Nesne dönüştürülemedi.".into(),
                Some("transform".into()),
            )));
        };
        sources.push((slot, uid.clone()));
        moved.push(e);
    }
    if sources.is_empty() {
        return Err(Stop::Failed(checks::error(
            codes::LAYER_LOCKED,
            locked_message(locked.len()),
            Some("uids".into()),
        )));
    }
    if overflow {
        return Err(Stop::Failed(checks::error(
            codes::NOT_FINITE,
            "Dönüşüm sonucunda sonlu olmayan bir değer çıktı (sayı taşması). Daha küçük bir değer verin.".into(),
            Some("transform".into()),
        )));
    }
    let mut warnings = Vec::new();
    if !locked.is_empty() {
        warnings.push(CommandWarning {
            code: codes::LAYER_LOCKED.into(),
            message: locked_message(locked.len()),
            path: Some("uids".into()),
        });
    }
    Ok(Checked {
        sources,
        moved,
        locked,
        warnings,
    })
}

/// Every number of a shape's geometry is finite (the fields the web's
/// handler walks: points, radii, angles, bulges, heights, offsets, the
/// hatch pattern's angle and spacing).
fn finite_shape(s: &Shape) -> bool {
    use kentos_geometry_core::Vec2;
    fn pt(p: &Vec2) -> bool {
        p.x.is_finite() && p.y.is_finite()
    }
    fn pts(ps: &[Vec2]) -> bool {
        ps.iter().all(pt)
    }
    fn values(vs: &Option<Vec<f64>>) -> bool {
        vs.iter().flatten().all(|v| v.is_finite())
    }
    match s {
        Shape::Point { p, z } => pt(p) && z.is_none_or(f64::is_finite),
        Shape::Line { a, b } => pt(a) && pt(b),
        Shape::Polyline {
            pts: p,
            bulges,
            holes,
        }
        | Shape::Polygon {
            pts: p,
            bulges,
            holes,
        } => {
            pts(p)
                && values(bulges)
                && holes
                    .iter()
                    .flatten()
                    .all(|h| pts(&h.pts) && values(&h.bulges))
        }
        Shape::Circle { c, r } => pt(c) && r.is_finite(),
        Shape::Arc { c, r, a0, a1 } => pt(c) && r.is_finite() && a0.is_finite() && a1.is_finite(),
        Shape::Ellipse {
            c,
            major,
            ratio,
            t0,
            t1,
        } => pt(c) && pt(major) && ratio.is_finite() && t0.is_finite() && t1.is_finite(),
        Shape::Xline { p, dir } | Shape::Ray { p, dir } => pt(p) && pt(dir),
        Shape::Spline { pts: p, .. } => pts(p),
        Shape::Text {
            p,
            height,
            rotation,
            ..
        } => pt(p) && height.is_finite() && rotation.is_finite(),
        Shape::Dimension {
            a,
            b,
            offset,
            height,
            angle,
            c,
            ..
        } => {
            pt(a)
                && pt(b)
                && offset.is_finite()
                && height.is_finite()
                && angle.is_none_or(f64::is_finite)
                && c.as_ref().is_none_or(pt)
        }
        Shape::Hatch {
            ring,
            holes,
            pattern,
        } => {
            pts(ring)
                && holes.iter().flatten().all(|h| pts(h))
                && pattern.angle.is_finite()
                && pattern.spacing.is_finite()
        }
    }
}

fn base_mut(e: &mut Entity) -> &mut kentos_contracts::EntityBase {
    match e {
        Entity::Point(e) => &mut e.base,
        Entity::Line(e) => &mut e.base,
        Entity::Polyline(e) | Entity::Polygon(e) => &mut e.base,
        Entity::Circle(e) => &mut e.base,
        Entity::Arc(e) => &mut e.base,
        Entity::Ellipse(e) => &mut e.base,
        Entity::Spline(e) => &mut e.base,
        Entity::Xline(e) | Entity::Ray(e) => &mut e.base,
        Entity::Text(e) => &mut e.base,
        Entity::Dimension(e) => &mut e.base,
        Entity::Hatch(e) => &mut e.base,
    }
}
